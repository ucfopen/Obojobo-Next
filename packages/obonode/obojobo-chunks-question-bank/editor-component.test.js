import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { Registry } from 'obojobo-document-engine/src/scripts/common/registry'
import QuestionBank from './editor-component'

import { Transforms } from 'slate'
jest.mock('slate')
import { ReactEditor } from 'slate-react'
jest.mock('slate-react')
jest.mock('./icon', () => global.mockReactComponent(this, 'Icon'))
jest.mock('./converter', () => ({ mock: 'converter' }))
jest.mock('obojobo-document-engine/src/scripts/common/registry', () => ({
	Registry: {
		registerModel: jest.fn(),
		getItemForType: jest.fn()
	}
}))
jest.mock(
	'obojobo-document-engine/src/scripts/oboeditor/components/node/with-slate-wrapper',
	() => item => item
)
jest.mock(
	'obojobo-document-engine/src/scripts/oboeditor/components/node/editor-component',
	() => props => <div>{props.children}</div>
)

jest.useFakeTimers()

describe('QuestionBank editor', () => {
	test('QuestionBank builds the expected component', () => {
		const props = {
			node: {
				data: {
					get: () => {
						return {}
					}
				}
			},
			parent: {
				getPath: () => ({
					get: () => 0
				}),
				nodes: {
					size: 2
				}
			},
			element: { content: {} }
		}

		const component = renderer.create(<QuestionBank {...props} />)
		const tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

	test('QuestionBank component changes choose type', () => {
		const props = {
			element: {
				content: { choose: 8, select: 'sequential' }
			},
			node: {
				key: 'mock_key'
			},
			parent: {
				getPath: () => ({
					get: () => 0
				}),
				nodes: {
					size: 2
				}
			}
		}

		const component = mount(<QuestionBank {...props} />)
		component
			.find('input')
			.at(0)
			.simulate('click')
		component
			.find('input')
			.at(0)
			.simulate('change', { target: { value: 'all' } })

		expect(component.html()).toMatchSnapshot()
	})

	test('QuestionBank component changes choose amount', () => {
		const props = {
			element: {
				content: { choose: 8, select: 'sequential' }
			},
			editor: {
				toggleEditable: jest.fn()
			},
			node: {
				key: 'mock_key'
			},
			parent: {
				getPath: () => ({
					get: () => 0
				}),
				nodes: {
					size: 2
				}
			}
		}

		const component = mount(<QuestionBank {...props} />)
		component
			.find('input')
			.at(1)
			.simulate('click')
		component
			.find('input')
			.at(1)
			.simulate('change', { target: { value: 'pick' } })

		component
			.find('input')
			.at(2)
			.simulate('focus')
		component
			.find('input')
			.at(2)
			.simulate('click')
		component
			.find('input')
			.at(2)
			.simulate('change', { target: { value: '7' } })
		component
			.find('input')
			.at(2)
			.simulate('blur')
		jest.runAllTimers()

		expect(component.html()).toMatchSnapshot()
	})

	test('QuestionBank component changes select type', () => {
		const props = {
			element: {
				content: { choose: 8, select: 'sequential' }
			},
			node: {
				key: 'mock_key'
			},
			parent: {
				getPath: () => ({
					get: () => 0
				}),
				nodes: {
					size: 2
				}
			}
		}

		const component = mount(<QuestionBank {...props} />)
		component
			.find('select')
			.at(0)
			.simulate('click')
		component
			.find('select')
			.at(0)
			.simulate('change', { target: { value: 'pick' } })

		expect(component.html()).toMatchSnapshot()
	})

	test('QuestionBank component deletes self', () => {
		Registry.getItemForType.mockReturnValueOnce({
			insertJSON: {
				type: 'Mock'
			}
		})

		const props = {
			node: {
				data: {
					data: {
						get: () => {
							return { choose: 8, select: 'sequential' }
						}
					}
				}
			},
			parent: {
				getPath: () => ({
					get: () => 0
				}),
				nodes: {
					size: 2
				}
			},
			editor: {
				removeNodeByKey: jest.fn()
			},
			element: {
				content: { choose: 8, select: 'sequential' }
			}
		}

		const component = mount(<QuestionBank {...props} />)

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(Transforms.removeNodes).toHaveBeenCalled()
	})

	test('QuestionBank component adds question', () => {
		Registry.getItemForType.mockReturnValueOnce({
			insertJSON: {
				type: 'Mock'
			}
		})

		const props = {
			element: {
				content: {},
				children: []
			},
			node: {
				key: 'mock_key'
			},
			parent: {
				getPath: () => ({
					get: () => 0
				}),
				nodes: {
					size: 2
				}
			},
			editor: {}
		}

		ReactEditor.findPath.mockReturnValueOnce([])

		const component = mount(<QuestionBank {...props} />)
		component
			.find('button')
			.at(1)
			.simulate('click')

		expect(Transforms.insertNodes).toHaveBeenCalledWith({}, { type: 'Mock' }, { at: [0] })
	})

	test('QuestionBank component adds question bank', () => {
		const props = {
			element: {
				content: {},
				children: []
			},
			node: {
				key: 'mock_key'
			},
			parent: {
				getPath: () => ({
					get: () => 0
				}),
				nodes: {
					size: 2
				}
			},
			editor: {}
		}

		ReactEditor.findPath.mockReturnValueOnce([])

		const component = mount(<QuestionBank {...props} />)
		component
			.find('button')
			.at(2)
			.simulate('click')

		expect(Transforms.insertNodes).toHaveBeenCalledWith(
			{},
			expect.objectContaining({ type: 'ObojoboDraft.Chunks.QuestionBank' }),
			{ at: [0] }
		)
	})

	test('QuestionBank component sets properties', () => {
		const props = {
			element: {
				content: {},
				children: []
			},
			editor: {},
			selected: true
		}

		ReactEditor.findPath.mockReturnValueOnce([])

		const component = mount(<QuestionBank {...props} />)

		component.setProps({ selected: false })
		expect(Transforms.setNodes).toHaveBeenCalled()
	})
})
