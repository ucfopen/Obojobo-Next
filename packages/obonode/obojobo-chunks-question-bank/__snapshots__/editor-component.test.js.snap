// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionBank editor QuestionBank builds the expected component 1`] = `
<div>
  <div
    className="obojobo-draft--chunks--question-bank editor-bank  is-open"
  >
    <div
      className="obojobo-draft--components--button is-not-dangerous align-center collapse-button"
    >
      <button
        className="button"
        contentEditable={false}
        onClick={[Function]}
      >
        ⌃
      </button>
    </div>
    <div
      className="obojobo-draft--components--button is-not-dangerous align-center delete-button"
    >
      <button
        className="button"
        contentEditable={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        ×
      </button>
    </div>
    <div
      className="question-bank-content"
    >
      <div
        className="qb-settings"
        contentEditable={false}
      >
        <fieldset
          className="choose"
        >
          <legend>
            How many questions should be displayed?
          </legend>
          <label>
            <input
              name="undefined-choose"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="radio"
              value="all"
            />
            All questions
          </label>
          <span>
             or
          </span>
          <label>
            <input
              checked={true}
              name="undefined-choose"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="radio"
              value="pick"
            />
            Pick
          </label>
          <input
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            type="number"
            value={1}
          />
        </fieldset>
        <label
          className="select"
        >
          How should questions be selected?
          <select
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            value="sequential"
          >
            <option
              value="sequential"
            >
              In order
            </option>
            <option
              value="random"
            >
              Randomly
            </option>
            <option
              value="random-unseen"
            >
              Randomly, with no repeats
            </option>
          </select>
        </label>
      </div>
      <div
        className="button-bar"
        contentEditable={false}
      >
        <div
          className="obojobo-draft--components--button is-not-dangerous align-center"
        >
          <button
            className="button"
            contentEditable={false}
            onClick={[Function]}
          >
            Add Question
          </button>
        </div>
        <div
          className="obojobo-draft--components--button is-not-dangerous align-center"
        >
          <button
            className="button"
            contentEditable={false}
            onClick={[Function]}
          >
            Add Question Bank
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`QuestionBank editor QuestionBank component changes choose amount 1`] = `"<div><div class=\\"obojobo-draft--chunks--question-bank editor-bank  is-open\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center collapse-button\\"><button class=\\"button\\" contenteditable=\\"false\\">⌃</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\" contenteditable=\\"false\\">×</button></div><div class=\\"question-bank-content\\"><div class=\\"qb-settings\\" contenteditable=\\"false\\"><fieldset class=\\"choose\\"><legend>How many questions should be displayed?</legend><label><input type=\\"radio\\" name=\\"undefined-choose\\" value=\\"all\\">All questions</label><span> or</span><label><input type=\\"radio\\" name=\\"undefined-choose\\" value=\\"pick\\" checked=\\"\\">Pick</label><input type=\\"number\\" value=\\"7\\"></fieldset><label class=\\"select\\">How should questions be selected?<select><option value=\\"sequential\\">In order</option><option value=\\"random\\">Randomly</option><option value=\\"random-unseen\\">Randomly, with no repeats</option></select></label></div><div class=\\"button-bar\\" contenteditable=\\"false\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center\\"><button class=\\"button\\" contenteditable=\\"false\\">Add Question</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center\\"><button class=\\"button\\" contenteditable=\\"false\\">Add Question Bank</button></div></div></div></div></div>"`;

exports[`QuestionBank editor QuestionBank component changes choose type 1`] = `"<div><div class=\\"obojobo-draft--chunks--question-bank editor-bank  is-open\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center collapse-button\\"><button class=\\"button\\" contenteditable=\\"false\\">⌃</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\" contenteditable=\\"false\\">×</button></div><div class=\\"question-bank-content\\"><div class=\\"qb-settings\\" contenteditable=\\"false\\"><fieldset class=\\"choose\\"><legend>How many questions should be displayed?</legend><label><input type=\\"radio\\" name=\\"undefined-choose\\" value=\\"all\\">All questions</label><span> or</span><label><input type=\\"radio\\" name=\\"undefined-choose\\" value=\\"pick\\" checked=\\"\\">Pick</label><input type=\\"number\\" value=\\"8\\" disabled=\\"\\"></fieldset><label class=\\"select\\">How should questions be selected?<select><option value=\\"sequential\\">In order</option><option value=\\"random\\">Randomly</option><option value=\\"random-unseen\\">Randomly, with no repeats</option></select></label></div><div class=\\"button-bar\\" contenteditable=\\"false\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center\\"><button class=\\"button\\" contenteditable=\\"false\\">Add Question</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center\\"><button class=\\"button\\" contenteditable=\\"false\\">Add Question Bank</button></div></div></div></div></div>"`;

exports[`QuestionBank editor QuestionBank component changes select type 1`] = `"<div><div class=\\"obojobo-draft--chunks--question-bank editor-bank  is-open\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center collapse-button\\"><button class=\\"button\\" contenteditable=\\"false\\">⌃</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\" contenteditable=\\"false\\">×</button></div><div class=\\"question-bank-content\\"><div class=\\"qb-settings\\" contenteditable=\\"false\\"><fieldset class=\\"choose\\"><legend>How many questions should be displayed?</legend><label><input type=\\"radio\\" name=\\"undefined-choose\\" value=\\"all\\">All questions</label><span> or</span><label><input type=\\"radio\\" name=\\"undefined-choose\\" value=\\"pick\\" checked=\\"\\">Pick</label><input type=\\"number\\" value=\\"8\\"></fieldset><label class=\\"select\\">How should questions be selected?<select><option value=\\"sequential\\">In order</option><option value=\\"random\\">Randomly</option><option value=\\"random-unseen\\">Randomly, with no repeats</option></select></label></div><div class=\\"button-bar\\" contenteditable=\\"false\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center\\"><button class=\\"button\\" contenteditable=\\"false\\">Add Question</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center\\"><button class=\\"button\\" contenteditable=\\"false\\">Add Question Bank</button></div></div></div></div></div>"`;
